halo2_proofs/src/arithmetic.rs

```diff
@@ -411,27 +397,105 @@ pub fn lagrange_interpolate<F: FieldExt>(points: &[F], evals: &[F]) -> Vec<F> {
 }
 
 #[cfg(test)]
-use rand_core::OsRng;
+mod tests {
+    use super::{
+        best_fft, best_ifft, eval_polynomial, lagrange_interpolate, swap_bit_reverse, Field, Group,
+    };
+    use crate::pasta::Fp;
+    use crate::poly::EvaluationDomain;
+    use proptest::prelude::*;
+    use rand_core::OsRng;
+
+    #[test]
+    fn test_bitreverse() {
+        fn bitreverse(mut n: usize, l: usize) -> usize {
+            let mut r = 0;
+            for _ in 0..l {
+                r = (r << 1) | (n & 1);
+                n >>= 1;
+            }
+            r
+        }
+        for k in 3..10 {
+            let n = 1 << k;
+            for i in 0..n as u64 {
+                assert_eq!(
+                    bitreverse(i as usize, k),
+                    (i.reverse_bits() >> (64 - k)) as usize
+                )
+            }
+        }
+    }
 
-#[cfg(test)]
-use crate::pasta::Fp;
+    fn prev_fft<G: Group>(a: &mut [G], omega: G::Scalar, log_n: u32) {
+        let n = a.len() as u32;
+        assert_eq!(n, 1 << log_n);
+
+        swap_bit_reverse(a, n as usize, log_n);
+
+        let mut m = 1;
+        for _ in 0..log_n {
+            let w_m = omega.pow_vartime(&[u64::from(n / (2 * m)), 0, 0, 0]);
+            let mut k = 0;
+            while k < n {
+                let mut w = G::Scalar::one();
+                for j in 0..m {
+                    let mut t = a[(k + j + m) as usize];
+                    t.group_scale(&w);
+                    a[(k + j + m) as usize] = a[(k + j) as usize];
+                    a[(k + j + m) as usize].group_sub(&t);
+                    a[(k + j) as usize].group_add(&t);
+                    w *= &w_m;
+                }
+                k += 2 * m;
+            }
+            m *= 2;
+        }
+    }
 
-#[test]
-fn test_lagrange_interpolate() {
-    let rng = OsRng;
+    proptest! {
+        #![proptest_config(ProptestConfig::with_cases(100))]
+        #[test]
+        fn test_fft(k in 3u32..10) {
```

Comment:
This proptest should be rewritten to take a single `k: u32`, and sample an arbitrary `a: [Fp: u32]` and `omega: Fp` 

```rust
fn test_fft(k: u32)(
    a in arb_poly(k),
    omega in arb_fp()
)
```
Then, we could run `test_fft(3)`, ..., `test_fft(10)` for the desired range of `k`.